lib/request.js

```diff
@@ -244,10 +245,29 @@ module.exports = {
   get hostname() {
     const host = this.host;
     if (!host) return '';
+    if ('[' == host[0]) return this.URL.hostname; // IPv6
     return host.split(':')[0];
   },
 
   /**
+   * Get WHATWG parsed URL.
+   * Lazily memoized.
+   *
+   * @return {URL}
+   * @api public
+   */
+
+  get URL() {
+    if (!this.memoizedURL) {
+      const protocol = this.protocol;
+      const host = this.host;
+      const originalUrl = this.originalUrl || ''; // avoid undefined in template string
+      this.memoizedURL = new URL(`${protocol}://${host}${originalUrl}`);
```

Comment:
URL will throw TypeError if something's off.
Koa is generally has `value || ''` type of solutions in most cases.

Maybe wrap this in a try-catch and memoize empty object so that `this.URL.hostname` when malformed does not throw/crash the entire app?